File: packages/shared/agent-loop.ts
export type AgentLoopState =
  | "start"
  | "call_model"
  | "check_api_response"
  | "accumulate_stream"
  | "classify_finish_reason"
  | "validate_tool_call"
  | "execute_tool"
  | "append_observation"
  | "checkpoint_if_needed"
  | "rebuild_context_if_needed"
  | "classify_content"
  | "goal_verify"
  | "final"
  | "think_only"
  | "api_error"
  | "truncated"
  | "filtered"
  | "failed";

export type AgentFinishReason =
  | "tool_calls"
  | "function_call"
  | "length"
  | "content_filter"
  | "stop"
  | "api_error"
  | "streaming_null"
  | "unknown";

export type AgentApiResponseStatus =
  | "ok"
  | "streaming"
  | "api_error"
  | "parse_error"
  | "incomplete";

export type AgentContentClassification =
  | "final_answer"
  | "think_only"
  | "empty"
  | "question";

export type AgentApiErrorClass =
  | "rate_limited"
  | "server_error"
  | "timeout"
  | "auth_error"
  | "client_error"
  | "parse_error"
  | "unknown";

export type AgentGoalVerdict = "complete" | "incomplete" | "impossible";

export interface AgentToolCallPayload {
  id?: string;
  name: string;
  arguments: string | Record<string, unknown>;
}

export interface AgentRetryPolicy {
  maxAttempts: number;
  retriableErrors: AgentApiErrorClass[];
}

export interface ClassifyFinishReasonInput {
  finishReason?: string | null;
  content?: string;
  toolCalls?: AgentToolCallPayload[];
  isStreaming?: boolean;
  receivedNullChunksOnly?: boolean;
}

export interface ClassifyContentInput {
  content: string;
  hadToolCallsInTurn?: boolean;
}

export type AgentLoopEventType =
  | "loop_started"
  | "model_call_dispatched"
  | "api_response_ok"
  | "api_response_streaming"
  | "api_response_error"
  | "stream_complete"
  | "stream_error"
  | "stream_null_stall"
  | "finish_classified"
  | "tool_valid"
  | "tool_invalid"
  | "tool_executed"
  | "observation_appended"
  | "checkpoint_done"
  | "rebuild_not_needed"
  | "rebuild_done"
  | "content_classified"
  | "goal_complete"
  | "goal_incomplete"
  | "goal_impossible"
  | "no_goal_set"
  | "nudge_sent"
  | "nudge_exhausted"
  | "retry_allowed"
  | "retry_exhausted"
  | "truncation_recovery_selected"
  | "truncation_terminal";

export interface AgentLoopTransition {
  from: AgentLoopState;
  to: AgentLoopState;
  event: AgentLoopEventType;
  reason?: string;
}

export interface AgentLoopDecision {
  nextState: AgentLoopState;
  reason: string;
  finishReason?: AgentFinishReason;
  contentClassification?: AgentContentClassification;
  terminal?: boolean;
  retryable?: boolean;
}

export interface AgentLoopEvent {
  type: AgentLoopEventType;
  finishReason?: AgentFinishReason;
  contentClassification?: AgentContentClassification;
  hasGoal?: boolean;
}

export const DEFAULT_AGENT_RETRY_POLICY: AgentRetryPolicy = {
  maxAttempts: 3,
  retriableErrors: ["rate_limited", "server_error", "timeout"],
};

export const TERMINAL_AGENT_LOOP_STATES: ReadonlySet<AgentLoopState> = new Set([
  "final",
  "failed",
  "filtered",
]);

File: crates/codra-protocol/src/agent_loop.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentLoopState {
    Start,
    CallModel,
    CheckApiResponse,
    AccumulateStream,
    ClassifyFinishReason,
    ValidateToolCall,
    ExecuteTool,
    AppendObservation,
    CheckpointIfNeeded,
    RebuildContextIfNeeded,
    ClassifyContent,
    GoalVerify,
    Final,
    ThinkOnly,
    ApiError,
    Truncated,
    Filtered,
    Failed,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentFinishReason {
    ToolCalls,
    FunctionCall,
    Length,
    ContentFilter,
    Stop,
    ApiError,
    StreamingNull,
    Unknown,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentApiResponseStatus {
    Ok,
    Streaming,
    ApiError,
    ParseError,
    Incomplete,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentContentClassification {
    FinalAnswer,
    ThinkOnly,
    Empty,
    Question,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentApiErrorClass {
    RateLimited,
    ServerError,
    Timeout,
    AuthError,
    ClientError,
    ParseError,
    Unknown,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentGoalVerdict {
    Complete,
    Incomplete,
    Impossible,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentLoopEventType {
    LoopStarted,
    ModelCallDispatched,
    ApiResponseOk,
    ApiResponseStreaming,
    ApiResponseError,
    StreamComplete,
    StreamError,
    StreamNullStall,
    FinishClassified,
    ToolValid,
    ToolInvalid,
    ToolExecuted,
    ObservationAppended,
    CheckpointDone,
    RebuildNotNeeded,
    RebuildDone,
    ContentClassified,
    GoalComplete,
    GoalIncomplete,
    GoalImpossible,
    NoGoalSet,
    NudgeSent,
    NudgeExhausted,
    RetryAllowed,
    RetryExhausted,
    TruncationRecoverySelected,
    TruncationTerminal,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AgentLoopTransition {
    pub from: AgentLoopState,
    pub to: AgentLoopState,
    pub event: AgentLoopEventType,
    pub reason: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AgentLoopDecision {
    pub next_state: AgentLoopState,
    pub reason: String,
    pub finish_reason: Option<AgentFinishReason>,
    pub content_classification: Option<AgentContentClassification>,
    pub terminal: Option<bool>,
    pub retryable: Option<bool>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn agent_loop_state_serializes_snake_case() {
        let state = AgentLoopState::ValidateToolCall;
        let json = serde_json::to_string(&state).expect("serialize state");
        assert_eq!(json, "\"validate_tool_call\"");
    }

    #[test]
    fn agent_finish_reason_serializes_snake_case() {
        let reason = AgentFinishReason::StreamingNull;
        let json = serde_json::to_string(&reason).expect("serialize finish reason");
        assert_eq!(json, "\"streaming_null\"");
    }

    #[test]
    fn agent_loop_decision_round_trips() {
        let decision = AgentLoopDecision {
            next_state: AgentLoopState::GoalVerify,
            reason: "goal required".to_string(),
            finish_reason: Some(AgentFinishReason::Stop),
            content_classification: Some(AgentContentClassification::FinalAnswer),
            terminal: Some(false),
            retryable: None,
        };

        let json = serde_json::to_string(&decision).expect("serialize decision");
        let parsed: AgentLoopDecision = serde_json::from_str(&json).expect("parse decision");
        assert_eq!(parsed.next_state, AgentLoopState::GoalVerify);
        assert_eq!(parsed.finish_reason, Some(AgentFinishReason::Stop));
    }
}